### PR TITLE
[Proposal] Break Threat Intel Indicator Match rules into Indicator-type rules

### PR DESCRIPTION
**Pull request by w0rk3r**
## Issues

https://github.com/elastic/security-team/issues/5673

## Summary

This PR proposes the following changes to the indicator match rules:

- Deprecate the two current rules
  - This will be done in two steps: This PR updates the rule title and guide to say that the rule is deprecated to 8.5+
  - After the rule release, we do a PR to deprecate them in the repo.
- Create four new indicator-type rules that support both Filebeat and TI integrations, which are min_stack(ed) to 8.5
- Adds support to process.hash.* and dll.hash.* fields for hash indicators
- Adds support to the url.domain field for URL indicators

### CODE REVIEW

File: rules/cross-platform/threat_intel_indicator_match_address_expiring.toml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+updated_date = "2023/05/22"
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a IP address indicator from the Anomali or Record future integrations match against a
```
Comment (approksiu) [inline rules/cross-platform/threat_intel_indicator_match_address_expiring.toml:13]: "an IP address"

File: rules/cross-platform/threat_intel_indicator_match_address.toml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+updated_date = "2023/05/22"
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a IP address indicator from the Threat Intel Filebeat module or integrations has a match
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_address.toml:13]:
```suggestion
This rule is triggered when an IP address indicator from the Threat Intel Filebeat module or integrations has a match
```

File: rules/cross-platform/threat_intel_indicator_match_address_expiring.toml (hunk 1 above)
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_address_expiring.toml:13]:
```suggestion
This rule is triggered when an IP address indicator from the Anomali or Record future integrations match against a
```

File: rules/cross-platform/threat_intel_indicator_match_hash.toml [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a hash indicator from the Threat Intel Filebeat module or integrations has a match against a
+event that contains file hashs, such as antivirus alerts, file operation events, etc.
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_hash.toml:14]:
```suggestion
This rule is triggered when a hash indicator from the Threat Intel Filebeat module or integrations has a match against an
event that contains file hashes, such as antivirus alerts, file operation events, etc.
```

File: rules/cross-platform/threat_intel_indicator_match_hash_expiring.toml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a hash indicator from the Anomali or Record future integrations match against a event that
+contains file hashs, such as antivirus alerts, file operation events, etc.
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_hash_expiring.toml:14]:
```suggestion
This rule is triggered when a hash indicator from the Anomali or Record future integrations match against an event that
contains file hashes, such as antivirus alerts, file operation events, etc.
```

File: rules/cross-platform/threat_intel_indicator_match_registry.toml [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a Windows registry indicator from the Threat Intel Filebeat module or integrations has a
+match against a event that contains registry data.
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_registry.toml:14]:
```suggestion
This rule is triggered when a Windows registry indicator from the Threat Intel Filebeat module or integrations has a
match against an event that contains registry data.
```

File: rules/cross-platform/threat_intel_indicator_match_registry_expiring.toml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+name = "Threat Intel Windows Registry Expiring Indicator Match"
+references = [
+    "https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-module-threatintel.html",
+    "https://www.elastic.co/guide/en/security/master/es-threat-intel-integrations.html",
+    "https://www.elastic.co/security/tip"
+]
+risk_score = 99
+rule_id = "2da6ae64-056a-45e3-aaf8-2aa09bbeadde"
+severity = "critical"
+tags = ["Elastic", "Windows", "Elastic Endgame", "Network"]
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_registry_expiring.toml:30]:
```suggestion
tags = ["Elastic", "Windows", "Elastic Endgame"]
```

File: rules/cross-platform/threat_intel_indicator_match_registry_expiring.toml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a Windows registry indicator from the Anomali or Record future integrations match against a
+event that contains registry data.
+"""
+from = "now-65m"
+index = ["auditbeat-*", "endgame-*", "filebeat-*", "logs-*", "packetbeat-*", "winlogbeat-*"]
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_registry_expiring.toml:17]:
```suggestion
index = ["auditbeat-*", "endgame-*", "filebeat-*", "logs-*", "winlogbeat-*"]
```

File: rules/cross-platform/threat_intel_indicator_match_registry.toml [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+"""
+from = "now-65m"
+index = ["auditbeat-*", "endgame-*", "filebeat-*", "logs-*", "packetbeat-*", "winlogbeat-*"]
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_registry.toml:17]:
```suggestion
index = ["auditbeat-*", "endgame-*", "filebeat-*", "logs-*", "winlogbeat-*"]
```

File: rules/cross-platform/threat_intel_indicator_match_registry.toml [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+name = "Threat Intel Windows Registry Indicator Match"
+references = [
+    "https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-module-threatintel.html",
+    "https://www.elastic.co/guide/en/security/master/es-threat-intel-integrations.html",
+    "https://www.elastic.co/security/tip"
+]
+risk_score = 99
+rule_id = "a61809f3-fb5b-465c-8bff-23a8a068ac60"
+severity = "critical"
+tags = ["Elastic", "Windows", "Elastic Endgame", "Network"]
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_registry.toml:30]:
```suggestion
tags = ["Elastic", "Windows", "Elastic Endgame"]
```

File: rules/cross-platform/threat_intel_indicator_match_url.toml [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a URL indicator from the Threat Intel Filebeat module or integrations has a match against a
+event that contains URL data, like DNS events, network logs, etc.
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_url.toml:14]:
```suggestion
This rule is triggered when a URL indicator from the Threat Intel Filebeat module or integrations has a match against an
event that contains URL data, like DNS events, network logs, etc.
```

File: rules/cross-platform/threat_intel_indicator_match_url_expiring.toml [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a URL indicator from the Anomali or Record future integrations match against a event that
+contains URL data, like DNS events, network logs, etc.
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_indicator_match_url_expiring.toml:14]:
```suggestion
This rule is triggered when a URL indicator from the Anomali or Record future integrations match against an event that
contains URL data, like DNS events, network logs, etc.
```

**Comment by terrancedejesus** [comment]
I’d vote to have a domain rule as well, those are important for C2 infrastructure and commonly referred to in TI because of commodity malware.

Email isn’t as big, but doesn’t hurt to have. I’d be curious how users would have email data to match that email address, maybe O365 integration?

File: rules/cross-platform/threat_intel_indicator_match_registry_expiring.toml [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+updated_date = "2023/05/22"
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a Windows registry indicator from the Anomali or Record future integrations match against an
```
Comment (terrancedejesus) [inline rules/cross-platform/threat_intel_indicator_match_registry_expiring.toml:13]:
```suggestion
This rule is triggered when a Windows registry indicator from the Anomali or Recorded Future integrations match against an
```

File: rules/cross-platform/threat_intel_indicator_match_address_expiring.toml [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+updated_date = "2023/05/22"
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when an IP address indicator from the Anomali or Record future integrations match against a
```
Comment (terrancedejesus) [inline rules/cross-platform/threat_intel_indicator_match_address_expiring.toml:13]:
```suggestion
This rule is triggered when an IP address indicator from the Anomali or Recorded Future integrations match against a
```

File: rules/cross-platform/threat_intel_indicator_match_hash_expiring.toml [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+updated_date = "2023/05/22"
+min_stack_comments = """
+Threat Intel Indicator Match general rules were deprecated in 8.7, adding these rules to 8.8 to avoid alert duplication
+"""
+min_stack_version = "8.8.0"
+
+[rule]
+author = ["Elastic"]
+description = """
+This rule is triggered when a hash indicator from the Anomali or Record future integrations match against an event that
```
Comment (terrancedejesus) [inline rules/cross-platform/threat_intel_indicator_match_hash_expiring.toml:13]:
```suggestion
This rule is triggered when a hash indicator from the Anomali or Recorded Future integrations match against an event that
```

File: rules/cross-platform/threat_intel_fleet_integrations.toml [hunk 15]
```diff
@@ -30,20 +32,41 @@ If an indicator matches a local observation, the following enriched fields will
...
+This rule was deprecated in the 8.8 version of the Elastic Stack for performance reasons and to avoid alert duplication due to the indicators expiration new feature. Users using 8.8+ versions should disable this rule and enable indicator-based rules instead:
+
+If using integrations that support indicator expiration (Anomali or Recorded Future): 
+
+* Threat Intel IP Address Active Indicator Match - 804df840-68b6-46f7-8b09-63ce14737a6f
+* Threat Intel Hash Active Indicator Match - 369ad21f-7d79-4f90-aa80-b456337ec82f
+* Threat Intel Windows Registry Active Indicator Match - 2da6ae64-056a-45e3-aaf8-2aa09bbeadde
+* Threat Intel URL Active Indicator Match - 47c61f6d-5f2d-41fc-be13-541a54ea1c3e
+
+If using other integrations or filebeat:
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_fleet_integrations.toml:64]:
```suggestion
```

File: tests/test_all_rules.py [hunk 16]
```diff
@@ -497,7 +497,15 @@ def test_integration_tag(self):
...
-                        "a198fbbd-9413-45ec-a269-47ae4ccf59ce"
+                        "a198fbbd-9413-45ec-a269-47ae4ccf59ce",
+                        "804df840-68b6-46f7-8b09-63ce14737a6f",
+                        "0c41e478-5263-4c69-8f9e-7dfd2c22da64",
+                        "369ad21f-7d79-4f90-aa80-b456337ec82f",
+                        "aab184d3-72b3-4639-b242-6597c99d8bca",
+                        "2da6ae64-056a-45e3-aaf8-2aa09bbeadde",
+                        "a61809f3-fb5b-465c-8bff-23a8a068ac60",
+                        "47c61f6d-5f2d-41fc-be13-541a54ea1c3e",
+                        "f3e22c8b-ea47-45d1-b502-b57b6de950b3"
```
Comment (w0rk3r) [inline tests/test_all_rules.py:508]:
```suggestion
                        "a198fbbd-9413-45ec-a269-47ae4ccf59ce",
                        "0c41e478-5263-4c69-8f9e-7dfd2c22da64",
                        "aab184d3-72b3-4639-b242-6597c99d8bca",
                        "a61809f3-fb5b-465c-8bff-23a8a068ac60",
                        "f3e22c8b-ea47-45d1-b502-b57b6de950b3"
```

File: rules/cross-platform/threat_intel_filebeat8x.toml [hunk 17]
```diff
@@ -30,20 +32,41 @@ If an indicator matches a local observation, the following enriched fields will
...
+
+If using integrations that support indicator expiration (Anomali or Recorded Future): 
+
+* Threat Intel IP Address Active Indicator Match - 804df840-68b6-46f7-8b09-63ce14737a6f
+* Threat Intel Hash Active Indicator Match - 369ad21f-7d79-4f90-aa80-b456337ec82f
+* Threat Intel Windows Registry Active Indicator Match - 2da6ae64-056a-45e3-aaf8-2aa09bbeadde
+* Threat Intel URL Active Indicator Match - 47c61f6d-5f2d-41fc-be13-541a54ea1c3e
+
+If using other integrations or filebeat:
+
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_filebeat8x.toml:65]:
```suggestion
```

File: rules/cross-platform/threat_intel_fleet_integrations.toml [hunk 18]
```diff
@@ -30,20 +32,33 @@ If an indicator matches a local observation, the following enriched fields will
...
+
 - If suspicious or malicious behavior is observed, take immediate action to isolate activity to prevent further post-compromise behavior.
 - One example of a response if a machine matched a command and control IP address would be to add an entry to a network device such as a firewall or proxy appliance to prevent any outbound activity from leaving that machine.
 - Another example of a response with a malicious file hash match would involve validating if the file was properly quarantined, reviewing current running processes for any abnormal activity, and investigating for any other follow-up actions such as persistence or lateral movement.
+
+## Setup
+
+This rule was deprecated in the 8.8 version of the Elastic Stack for performance reasons and to avoid alert duplication due to the indicators expiration new feature. Users using 8.8+ versions should disable this rule and enable indicator-based rules instead:
+
+
```
Comment (w0rk3r) [inline rules/cross-platform/threat_intel_fleet_integrations.toml:57]:
```suggestion
```

**Comment by MakoWish** [comment]
> I’d vote to have a domain rule as well, those are important for C2 infrastructure and commonly referred to in TI because of commodity malware.

I agree. Whenever I get an external notice of a campaign involving software we use, I always search the related indices for the IoC's, and they very frequently include domains. 

> I’d be curious how users would have email data to match that email address, maybe O365 integration?

We use O365, but I don't know that it includes email events. I can't seem to find any. We also use Digital Guardian (DG) for DLP and Mimecast for email protection, and both include email events. There is a Mimecast integration for Elastic Agent (we are using), but not one yet for DG (currently using API to poll DG events and Logstash to parse them to ECS).

**Comment by w0rk3r** [comment]
> I agree. Whenever I get an external notice of a campaign involving software we use, I always search the related indices for the IoC's, and they very frequently include domains.

@MakoWish I just added this to the `Threat Intel URL Indicator Match` one if you want to review it ;)

> We use O365, but I don't know that it includes email events. I can't seem to find any. We also use Digital Guardian (DG) for DLP and Mimecast for email protection, and both include email events. There is a Mimecast integration for Elastic Agent (we are using), but not one yet for DG (currently using API to poll DG events and Logstash to parse them to ECS).

If you want, it would be awesome if you could open an issue with some redacted logs so we can get to know which fields we could match against, and I could develop something in the next iteration

**Review by DefSecSentinel — APPROVED** [review]
LGTM

**Comment by MakoWish** [comment]
> @MakoWish I just added this to the `Threat Intel URL Indicator Match` one if you want to review it ;)

@terrancedejesus, not sure what your thoughts were, but I was actually thinking the domain rule would be matches against `source.domain` or `destination.domain`. Including `url.domain` is not a bad idea, however.

**Pull request merged**